### PR TITLE
Detect the OS and switch node runner if not supported for Node20

### PR DESCRIPTION
**Pull request by qianz2**
Detect which operating system the agent is running on. 
If the platform is not a supported version, then use node16 Nodehandler instead of node20 Nodehandler.
Add deprecated warnings for those incompatible platforms such as ubuntu 18.04.
Add a knob UseNode20InUnsupportedSystem to force use node20 in an unsupported system.

https://github.com/nodesource/distributions

### CODE REVIEW

**Review by max-zaytsev — APPROVED** [review]

File: src/Agent.Sdk/Util/PlatformUtil.cs [hunk 1]
```diff
@@ -124,29 +121,29 @@ public static SystemVersion GetSystemVersion()
...
-        private static void DetectRHEL6()
+        private static void DetectRHELVersion(string version)
         {
-            lock (detectedRHEL6lock)
+            lock (detectedRHELVersionlock)
             {
                 if (!RunningOnLinux || !File.Exists("/etc/redhat-release"))
                 {
-                    detectedRHEL6 = false;
+                    detectedRHELVersion = false;
```
Comment (merlynomsft) [inline src/Agent.Sdk/Util/PlatformUtil.cs:130]: >detectedRHELVersion [](http://example.com/codeflow?start=20&length=19)

[maybe not for this PR], having this function change static members as a side effect not a good design, we should consider refactoring

File: src/Agent.Worker/Handlers/NodeHandler.cs [hunk 2]
```diff
@@ -333,6 +345,28 @@ private void OnDataReceived(object sender, ProcessDataReceivedEventArgs e)
...
+                Trace.Info($"Detected Ubuntu version <= 18.04, using node16 as execution handler, instead node20");
+                return false;
+            }
+            if (systemName.Equals("debian") &&
+                int.TryParse(systemVersion, out int debianVersion) &&
+                debianVersion <= 9) {
+                Trace.Info($"Detected Debian version <= 9, using node16 as execution handler, instead node20");
+                return false;
+            } 
+            if (PlatformUtil.RunningOnRHELVersion("6") || PlatformUtil.RunningOnRHELVersion("7")) {
```
Comment (merlynomsft) [inline src/Agent.Worker/Handlers/NodeHandler.cs:363]: >("6" [](http://example.com/codeflow?start=49&length=4)

should be "6." or "7." to match existing logic in platformutil

File: src/Agent.Worker/Handlers/NodeHandler.cs [hunk 3]
```diff
@@ -333,6 +345,28 @@ private void OnDataReceived(object sender, ProcessDataReceivedEventArgs e)
             }
         }
 
+        private bool isNode20SupportedSystems() {
```
Comment (merlynomsft) [inline src/Agent.Worker/Handlers/NodeHandler.cs:348]: >isNode20SupportedSystems [](http://example.com/codeflow?start=21&length=24)

convention is PascalCase here

Comment (merlynomsft) [inline src/Agent.Worker/Handlers/NodeHandler.cs:348]: >bool [](http://example.com/codeflow?start=16&length=4)

suggest collecting telemetry on detected OS / version for validation

File: src/Agent.Worker/DiagnosticLogManager.cs [hunk 4]
```diff
@@ -183,7 +183,7 @@ public sealed class DiagnosticLogManager : AgentService, IDiagnosticLogManager
             }
 
             bool dumpPackagesVerificationResult = AgentKnobs.DumpPackagesVerificationResult.GetValue(executionContext).AsBoolean();
-            if (dumpPackagesVerificationResult && PlatformUtil.RunningOnLinux && !PlatformUtil.RunningOnRHEL6)
+            if (dumpPackagesVerificationResult && PlatformUtil.RunningOnLinux && !PlatformUtil.RunningOnRHELVersion("6"))
```
Comment (merlynomsft) [inline src/Agent.Worker/DiagnosticLogManager.cs:186]: >("6") [](http://example.com/codeflow?start=115&length=5)

should be "6." to match expected logic; I'd suggest making a static field or encapsulating (E.g. exposing RunningOnRHELVersion6 / RunningOnRHELVersion7)

File: src/Agent.Worker/Handlers/NodeHandler.cs [hunk 5]
```diff
@@ -233,7 +233,14 @@ public string GetNodeLocation()
 
             string nodeFolder = NodeHandler.nodeFolder;
 
-            if (taskHasNode20Data)
+            if (taskHasNode20Data && !isNode20SupportedSystems())
+            {
+                Trace.Warning($"The operating system the agent is running on doesn't support Node20. " +
+                             "Please upgrade the operating system of this host to ensure compatibility with Node20 tasks: " +
```
Comment (merlynomsft) [inline src/Agent.Worker/Handlers/NodeHandler.cs:239]: should be localized (put in resource file) -- also please post a screenshot of the message (or at least the text) to the internal channel and tag @geekzter for review

I think the message should mention that we're using the node16 runner instead

**Review by merlynomsft — APPROVED** [review]
:shipit:

File: src/Agent.Worker/Handlers/NodeHandler.cs [hunk 6]
```diff
@@ -233,7 +233,14 @@ public string GetNodeLocation()
 
             string nodeFolder = NodeHandler.nodeFolder;
 
-            if (taskHasNode20Data)
+            if (taskHasNode20Data && !isNode20SupportedSystems())
```
Comment (merlynomsft) [inline src/Agent.Worker/Handlers/NodeHandler.cs:236]: >isNode20SupportedSystems [](http://example.com/codeflow?start=38&length=24)

would be good to add a knob for this to 'force' allowing tasks to run on node20 runner of "isNode20SupportedSystems"

File: src/Agent.Listener/node20.json [hunk 7]
```diff
@@ -0,0 +1,46 @@
+[
+    {
+      "id": "amzn",
+      "versions": [
```
[the remark this answers is not in the document]
Comment (qianz2) [reply src/Agent.Listener/node20.json:4]: It listed out all supported system for node20, used for checking if a system supported node20 after deserializing it. I removed this feature so deleting this file. Thanks for catching it.

File: src/Agent.Worker/Handlers/NodeHandler.cs (hunk 3 above)
Comment (qianz2) [reply src/Agent.Worker/Handlers/NodeHandler.cs]: updated

File: src/Agent.Worker/DiagnosticLogManager.cs (hunk 4 above)
Comment (qianz2) [reply src/Agent.Worker/DiagnosticLogManager.cs]: Updated. I added back static field RunningOnRHEL6 and RunningOnRHEL7.

File: src/Agent.Worker/Handlers/NodeHandler.cs (hunk 2 above)
Comment (qianz2) [reply src/Agent.Worker/Handlers/NodeHandler.cs]: updated

File: src/Agent.Worker/Handlers/NodeHandler.cs (hunk 6 above)
Comment (qianz2) [reply src/Agent.Worker/Handlers/NodeHandler.cs]: We already have one to force agent to use node16: AGENT_USE_NODE16, this one might already serve for this purpose?
https://github.com/microsoft/azure-pipelines-agent/blob/a3067b75926acec16ef0166ca34c3f6ad4e1ba82/src/Agent.Sdk/Knob/AgentKnobs.cs#L148C1-L153C52

File: src/Agent.Worker/Handlers/NodeHandler.cs (hunk 3 above)
Comment (qianz2) [reply src/Agent.Worker/Handlers/NodeHandler.cs]: Updated: added Trace.Info() for detected system os/ verison.

**Pull request merged**